### PR TITLE
change sequencing for parent-search

### PR DESCRIPTION
**Pull request by ZacLN**

### CODE REVIEW

File: src/requests/textdocument.jl [hunk 1]
```diff
@@ -8,22 +8,17 @@ function process(r::JSONRPC.Request{Val{Symbol("textDocument/didOpen")},DidOpenT
         doc._version = r.params.textDocument.version
         get_line_offsets(doc, true)
```
Comment (davidanthoff) [inline src/requests/textdocument.jl]: This branch needs a `set_open_in_editor(doc, true)`.

Also, doesn't this branch also need a call to `try_to_load_parents`?

File: src/requests/textdocument.jl [hunk 2]
```diff
@@ -8,22 +8,17 @@ function process(r::JSONRPC.Request{Val{Symbol("textDocument/didOpen")},DidOpenT
...
-            set_open_in_editor(doc, true)
-            if is_ignored(uri, server)
-                doc._runlinter = false
-            end
+        doc = Document(uri, r.params.textDocument.text, false, server)
+        setdocument!(server, URI2(uri), doc)
+        doc._version = r.params.textDocument.version
+        if any(i->startswith(uri, filepath2uri(i)), server.workspaceFolders)
+            doc._workspace_file = true
         end
```
Comment (davidanthoff) [inline src/requests/textdocument.jl]: `doc._workspace_file = any(i->startswith(uri, filepath2uri(i)), server.workspaceFolders)`

File: src/requests/textdocument.jl [hunk 3]
```diff
@@ -8,22 +8,17 @@ function process(r::JSONRPC.Request{Val{Symbol("textDocument/didOpen")},DidOpenT
...
+        set_open_in_editor(doc, true)
+        if is_ignored(uri, server)
+            doc._runlinter = false
+        end
```
Comment (davidanthoff) [inline src/requests/textdocument.jl]: `doc._runlinter = !is_ignored(uri, server)`

**Comment by ZacLN** [comment]
No, an edit within any given document wont't impact/change whether a parent file that includes it. What's your thinking?

File: src/requests/textdocument.jl [hunk 4]
```diff
@@ -59,10 +52,20 @@ function process(r::JSONRPC.Request{Val{Symbol("textDocument/didClose")},DidClos
     doc = getdocument(server, URI2(uri))
     empty!(doc.diagnostics)
     publish_diagnostics(doc, server)
-    if !is_workspace_file(doc)
-        deletedocument!(server, URI2(uri))
-    else
+    if is_workspace_file(doc)
         set_open_in_editor(doc, false)
+    else
+        if any(d.root == doc.root && doc._open_in_editor for (uri, d::Document) in getdocuments_pair(server) if d != doc)
```
Comment (davidanthoff) [inline src/requests/textdocument.jl]:
```suggestion
        if any(d.root == doc.root for d::Document in getdocuments_value(server) if d != doc)
```
I think we don't need to check for `doc._open_in_editor`, we will only get this message for docs that _are_ open in the editor by definition, right?

File: src/requests/textdocument.jl [hunk 5]
```diff
@@ -59,10 +52,20 @@ function process(r::JSONRPC.Request{Val{Symbol("textDocument/didClose")},DidClos
...
+            # If any other open document shares doc's root we just mark it as closed...
+            set_open_in_editor(doc, false)
+        else
+            # ...otherwise we delete all documents that share root with doc.
+            for (u,d) in getdocuments_pair(server)
+                if d.root == doc.root
```
Comment (davidanthoff) [inline src/requests/textdocument.jl]:
```suggestion
                if d.root == doc.root && !get_open_in_editor(d)
```
We need to make sure that we don't delete files here that are outside the workspace but are currently also open in the editor, I believe? If you accept this change here, you also need to add a `set_open_in_editor(doc, false)` before this loop, otherwise the doc that is actually being close won't get removed.

In general this seems all very tricky and hard to follow logic... if we do this here in `didClose`, do we need to do a similar thing in delete messages from file watch? I think this would all be way, way simpler if we went with https://github.com/julia-vscode/LanguageServer.jl/issues/544 :) I at least have a really hard time figuring out all the possible scenarios/cases here...

File: src/requests/textdocument.jl (hunk 4 above)
Comment (ZacLN) [reply src/requests/textdocument.jl]: Ooops, this is a typo, should read `d._open_in_editor` which is actually consistent with my comment within the file...

Changing the above addresses your comment below

**Comment by davidanthoff** [comment]
> No, an edit within any given document wont't impact/change whether a parent file that includes it. What's your thinking?

Ah, I didn't get the whole logic here, i.e. that this was looking for other files that are including the file that is being opened. Agreed then.
